[silicon_creator] Ensure OTBN is idle before writes/commands.

jadephilipoom opened the pull request:
Adjust the silicon_creator OTBN driver to check that OTBN is idle before all writes/commands, and wait if not. Since OTBN now performs a secure wipe after coming out of reset, this check is needed to ensure mask ROM code waits for the secure wipe to finish before loading or running applications.

See https://github.com/lowRISC/opentitan/issues/13753 for some discussion.

CC @GregAC @johannheyszl @alphan

alphan commented on sw/device/silicon_creator/lib/drivers/otbn.c:
```diff
@@ -59,6 +59,17 @@ static rom_error_t check_offset_len(uint32_t offset_bytes, size_t num_words,
   return kErrorOk;
 }
 
+/**
```
We already have `otbn_is_busy()`. WDYT about leaving the driver as is and adding
```
while(otbn_is_busy()) {
}
```
before we start using OTBN.

Otherwise we will be checking both before and after a command, which is a bit redundant.

jadephilipoom replied:
Sure, that seems fine. Also probably worth adding a basic check that makes sure OTBN isn't locked -- that way by process of elimination it must be idle. And an error is better than an infinite spin in the locked case (although I think OTBN should not be locked here unless it detects an FI attack).

jadephilipoom replied:
Updated the PR to add the `otbn_is_busy()` loop and the locked check to `otbn_cmd_run`, `otbn_write`, and `otbn_dmem_read` (since no writes/reads are permitted in non-idle states). I also added some unit tests for locked states.

alphan replied:
> OTBN should not be locked here unless it detects an FI attack

Thanks for updating the PR. I agree, I was thinking about the same thing. It looks like the driver also needs something similar to `otbn_busy_wait_for_done()` which is currently in `otbn_util.c`. WDYT about replacing `otbn_is_busy()` with something like:
```
rom_error_t otbn_busy_wait_for_done(void) {
    uint32_t status = UINT32_MAX;
    rom_error_t res = kErrorOk ^ status;
    do {
        status = abs_mmio_read32(kBase + OTBN_STATUS_REG_OFFSET);
    } while (launder(status) != kOtbnStatusIdle && launder(status) != kOtbnStatusLocked);
    res ^= ~status;
    if (launder(res) == kErrorOk) {
        HARDENED_CHECK_EQ(res, kErrorOk);
        HARDENED_CHECK_EQ(abs_mmio_read32(kBase + OTBN_STATUS_REG_OFFSET), kOtbnStatusIdle);
        return res;
    }
    return kOtbnUnavailable;
}
```

If this makes sense, we can update the call sites that use `otbn_is_busy()` and #13931.

jadephilipoom replied:
I've replaced `otbn_is_busy()` with this code, removed the blocking parts from `otbn_write` and `otbn_cmd_run`, and added an `otbn_busy_wait_for_done()` step in `otbn_load_app` (so it runs before mask ROM loads the modexp application but not for subsequent operations).

Just to check that this is the intended outcome: if OTBN is locked, this will produce a hardening check failure. Since OTBN being locked means we should suspect a fault injection attack, this makes sense to me.

alphan commented on sw/device/silicon_creator/lib/drivers/otbn.c:
```diff
@@ -59,15 +59,32 @@ static rom_error_t check_offset_len(uint32_t offset_bytes, size_t num_words,
-                       size_t num_words) {
+static rom_error_t otbn_write(uint32_t dest_addr, const uint32_t *src,
+                              size_t num_words) {
+  // If OTBN is busy, wait for it to be done.
```
nit: I think it's enough to perform this check once before starting to use OTBN, e.g. in `mod_exp_otbn.c`

alphan commented on sw/device/silicon_creator/lib/drivers/otbn.c:
```diff
@@ -89,7 +108,6 @@ static void otbn_write(uint32_t dest_addr, const uint32_t *src,
  * This function blocks until OTBN is idle.
  *
  * @param cmd OTBN command.
- * @param exp_status Expected OTBN status after issuing the command.
```
Thanks for cleaning this up!

alphan reviewed (approved):
Thanks a lot, lgtm! Can you upload the disassembly to double check the generated code?

jadephilipoom commented:
Disassembly is [here](https://gist.github.com/jadephilipoom/ccf9406b3c27263c66bfa4adf57fe45f#file-rom_fpga_cw310-elf)

The new call to `otbn_busy_wait_for_done` from within `otbn_load_app` is on  [line 10288](https://gist.github.com/jadephilipoom/ccf9406b3c27263c66bfa4adf57fe45f#file-rom_fpga_cw310-elf-s-L10288). The body of `otbn_busy_wait_for_done` starts on [line 10496](https://gist.github.com/jadephilipoom/ccf9406b3c27263c66bfa4adf57fe45f#file-rom_fpga_cw310-elf-s-L10496) and is annotated below.

### Summary

The compiler slightly optimized the computation of `res` and `status`, but not in a way that seems dangerous to me. The original source says:
```c
rom_error_t otbn_busy_wait_for_done(void) {
  uint32_t status = UINT32_MAX;
  rom_error_t res = kErrorOk ^ status;
  do {
    status = abs_mmio_read32(kBase + OTBN_STATUS_REG_OFFSET);
  } while (launder32(status) != kOtbnStatusIdle &&
           launder32(status) != kOtbnStatusLocked);
  res ^= ~status;
  if (launder32(res) == kErrorOk) {
...
```
...while the compiled code does something more like:
```c
rom_error_t otbn_busy_wait_for_done(void) {
  do {
    status = abs_mmio_read32(kBase + OTBN_STATUS_REG_OFFSET);
  } while (launder32(status) != kOtbnStatusIdle &&
           launder32(status) != kOtbnStatusLocked);
  rom_errot_t res = kErrorOk ^ status;
  if (launder32(res) == kErrorOk) {
...
```
It seems to me that the compiler has simply turned `kErrorOk ^ UINT32_MAX ^ ~status` into `kErrorOk ^ status`. @alphan does this look OK to you?

### Detailed annotation

Set `a0` to `TOP_EARLGREY_OTBN_BASE_ADDR` (which is `0x41130`)
```asm
rom_error_t otbn_busy_wait_for_done(void) {
    c138:              41130537                 lui     a0,0x41130
```
Set `a1` to `kOtbnStatusLocked` (which is `255`).
```
    c13c:              0ff00593                 li      a1,255
```
Top of do/while loop: `a2` gets the result of reading OTBN's status register (`OTBN_STATUS_REG_OFFSET` = `0x18` = 24)
```asm
abs_mmio_read32():
/proc/self/cwd/./sw/device/lib/base/abs_mmio.h:73
  return *((volatile uint32_t *)addr);
    c140:          /-> 4d10                     lw      a2,24(a0)
```
Copy `a2` to `a3` (this is an artifact of `launder32`).
```asm
launder32():
/proc/self/cwd/./sw/device/lib/base/hardened.h:243
    c142:          |   86b2                     mv      a3,a2
```
If `a3` is zero (i.e. `status = kOtbnStatusIdle`), then exit the loop by jumping to `c14c`.
```asm
otbn_busy_wait_for_done():
/proc/self/cwd/sw/device/silicon_creator/lib/drivers/otbn.c:67
  uint32_t status = UINT32_MAX;
  rom_error_t res = kErrorOk ^ status;
  do {
    status = abs_mmio_read32(kBase + OTBN_STATUS_REG_OFFSET);
  } while (launder32(status) != kOtbnStatusIdle &&
    c144:       /--|-- c681                     beqz    a3,c14c <otbn_busy_wait_for_done+0x14>
```
Copy `a2` to `a3` (again, an artifact of `launder32`).
```asm
launder32():
/proc/self/cwd/./sw/device/lib/base/hardened.h:243
    c146:       |  |   86b2                     mv      a3,a2
```
Branch back to the start of the loop (`c140`) if `a3 != a1` (i.e. if `status != kOtbnStatusLocked`).
```asm
otbn_busy_wait_for_done():
/proc/self/cwd/sw/device/silicon_creator/lib/drivers/otbn.c:67
    c148:       |  \-- feb69ce3                 bne     a3,a1,c140 <otbn_busy_wait_for_done+0x8>
```
Set `a0` to the XOR of `status` (stored in `a2`) and `kErrorOk` (`1849` = `0x739`). 
```asm
/proc/self/cwd/sw/device/silicon_creator/lib/drivers/otbn.c:69
           launder32(status) != kOtbnStatusLocked);
  res ^= ~status;
    c14c:       \----> 73964513                 xori    a0,a2,1849
```
Copy `a0` to `a2` becaue of `launder32` and set `a1` to `kErrorOk` (`1849` = `0x739`).
```asm
launder32():
/proc/self/cwd/./sw/device/lib/base/hardened.h:243
    c150:              862a                     mv      a2,a0
    c152:              73900593                 li      a1,1849
```
Branch to `c178` (`return kErrorOtbnUnavailable`) if `a2 != a1` (i.e. `res != kErrorOk`).
```asm
otbn_busy_wait_for_done():
/proc/self/cwd/sw/device/silicon_creator/lib/drivers/otbn.c:70
  if (launder32(res) == kErrorOk) {
    c156:       /----- 02b61163                 bne     a2,a1,c178 <otbn_busy_wait_for_done+0x40>
```
Branch to `c166` if `a0 == a1` (i.e. `res == kErrorOk`). Otherwise, execute `unimp` (hardened check failure).
```asm
/proc/self/cwd/sw/device/silicon_creator/lib/drivers/otbn.c:71
    HARDENED_CHECK_EQ(res, kErrorOk);
    c15a:       |  /-- 00b50663                 beq     a0,a1,c166 <otbn_busy_wait_for_done+0x2e>
    c15e:       |  |   0000                     unimp
    c160:       |  |   0000                     unimp
    c162:       |  |   0000                     unimp
    c164:       |  |   0000                     unimp
```
Load  the status register contents into `a1` (`0x41130 = TOP_EARLGREY_OTBN_BASE_ADDR` and `24 = OTBN_STATUS_REG_OFFSET`).
```asm
    c166:       |  \-> 411305b7                 lui     a1,0x41130
abs_mmio_read32():
/proc/self/cwd/./sw/device/lib/base/abs_mmio.h:73
    c16a:       |      4d8c                     lw      a1,24(a1)
```
Branch to `c176` (`return res`) if `a1 == 0` (that is, the current status is still idle). Otherwise, execute `unimp` because the hardening check failed.
```asm
otbn_busy_wait_for_done():
/proc/self/cwd/sw/device/silicon_creator/lib/drivers/otbn.c:72
    HARDENED_CHECK_EQ(abs_mmio_read32(kBase + OTBN_STATUS_REG_OFFSET),
    c16c:       |  /-- c589                     beqz    a1,c176 <otbn_busy_wait_for_done+0x3e>
    c16e:       |  |   0000                     unimp
    c170:       |  |   0000                     unimp
    c172:       |  |   0000                     unimp
    c174:       |  |   0000                     unimp
```
Return `res` (successful case.
```asm
/proc/self/cwd/sw/device/silicon_creator/lib/drivers/otbn.c:77
                      kOtbnStatusIdle);
    return res;
  }
  return kErrorOtbnUnavailable;
}
    c176:       |  \-> 8082                     ret
```
Return `kErrorOtbnUnavailable` (unsuccessful case).
```asm
    c178:       \----> 06425537                 lui     a0,0x6425
    c17c:              e0d50513                 addi    a0,a0,-499 # 6424e0d <_epmp_stack_guard_na4+0x241d60d>
    c180:              8082                     ret
```

jadephilipoom commented:
(Latest push just rebases on master so the diff here is up to date with the disassembly.)

alphan commented on sw/device/silicon_creator/lib/drivers/otbn.c:
```diff
@@ -59,6 +59,23 @@ static rom_error_t check_offset_len(uint32_t offset_bytes, size_t num_words,
 }
 
+rom_error_t otbn_busy_wait_for_done(void) {
+  uint32_t status = UINT32_MAX;
```
Can you give this a try as we discussed offline?

```suggestion
  uint32_t status = launder32(UINT32_MAX);
```

alphan commented:
Thanks for the detailed and clear analysis @jadephilipoom! Great job!

jadephilipoom commented:
In the latest push I fixed the compiler optimization issue by adding two `launder32()`s:
```diff
@@ -60,8 +60,8 @@ static rom_error_t check_offset_len(uint32_t offset_bytes, size_t num_words,
 }
 
 rom_error_t otbn_busy_wait_for_done(void) {
-  uint32_t status = UINT32_MAX;
-  rom_error_t res = kErrorOk ^ status;
+  uint32_t status = launder32(UINT32_MAX);
+  rom_error_t res = launder32(kErrorOk ^ status);
   do {
     status = abs_mmio_read32(kBase + OTBN_STATUS_REG_OFFSET);
   } while (launder32(status) != kOtbnStatusIdle &&
```

It turned out that I needed both to get the compiler to put the `xor` before the loop. Even then, the compiler doesn't choose the same register for `status = UINT32_MAX` and `status = abs_mmio_read32(...)`. But I think this is still an improvement over the previous version.

Updated disassembly is [here](https://gist.github.com/jadephilipoom/d059317e597f8570db96398246e90b7d). Here's the relevant bit annotated (starting at [line 10496](https://gist.github.com/jadephilipoom/d059317e597f8570db96398246e90b7d#file-rom_fpga_cw310-elf-s-L10496)).

Set `a0` to `UINT32_MAX ^ kErrorOk`.
```asm
rom_error_t otbn_busy_wait_for_done(void) {
    c138:              557d                     li      a0,-1
/proc/self/cwd/sw/device/silicon_creator/lib/drivers/otbn.c:64
  uint32_t status = launder32(UINT32_MAX);
  rom_error_t res = launder32(kErrorOk ^ status);
    c13a:              73954513                 xori    a0,a0,1849
```

Set `a1` to `TOP_EARLGREY_OTBN_BASE_ADDR` and `a2` to `kOtbnStatusLocked`.
```asm
    c13e:              411305b7                 lui     a1,0x41130
    c142:              0ff00613                 li      a2,255
```

Read OTBN status register and store the result in `a3`.
```
abs_mmio_read32():
/proc/self/cwd/./sw/device/lib/base/abs_mmio.h:73
  return *((volatile uint32_t *)addr);
    c146:          /-> 4d94                     lw      a3,24(a1)
```
Copy `a3` to `a4` (artifact of `launder32()`).
```asm
launder32():
/proc/self/cwd/./sw/device/lib/base/hardened.h:243
    c148:          |   8736                     mv      a4,a3
```
Exit the loop if `a4` is `kOtbnStatusIdle` (which happens to be 0, so we can use `beqz`).
```asm
otbn_busy_wait_for_done():
/proc/self/cwd/sw/device/silicon_creator/lib/drivers/otbn.c:67
  do {
    status = abs_mmio_read32(kBase + OTBN_STATUS_REG_OFFSET);
  } while (launder32(status) != kOtbnStatusIdle &&
    c14a:       /--|-- c701                     beqz    a4,c152 <otbn_busy_wait_for_done+0x1a>
```
Another `launder32()` copy.
```asm
launder32():
/proc/self/cwd/./sw/device/lib/base/hardened.h:243
    c14c:       |  |   8736                     mv      a4,a3
```
Branch back to the beginning of the loop unless `a4` is `kOtbnStatusLocked`.
```asm
otbn_busy_wait_for_done():
/proc/self/cwd/sw/device/silicon_creator/lib/drivers/otbn.c:67
    c14e:       |  \-- fec71ce3                 bne     a4,a2,c146 <otbn_busy_wait_for_done+0xe>
```
End of loop: set `x0` (`res`) to `x0 ^ a3` (`res ^ status`).
```asm
/proc/self/cwd/sw/device/silicon_creator/lib/drivers/otbn.c:69
           launder32(status) != kOtbnStatusLocked);
  res ^= ~status;
    c152:       \----> 8d35                     xor     a0,a0,a3
```
Invert `res` (note: here the compiler has changed `res ^= ~status` into `res = ~(res ^ status)`.
```
    c154:              fff54513                 not     a0,a0
```
From here, the procedure is pretty much the same as it was before.
```
launder32():
/proc/self/cwd/./sw/device/lib/base/hardened.h:243
    c158:              862a                     mv      a2,a0
    c15a:              73900593                 li      a1,1849
otbn_busy_wait_for_done():
/proc/self/cwd/sw/device/silicon_creator/lib/drivers/otbn.c:70
  if (launder32(res) == kErrorOk) {
    c15e:       /----- 02b61163                 bne     a2,a1,c180 <otbn_busy_wait_for_done+0x48>
...
```
